Remove service update name flag

thaJeztah opened the pull request:
The `--name` flag was inadvertently added to `docker service update`, but is not supported, as it has various side-effects (e.g., existing tasks are not renamed).

This removes the flag from the service update command and API documentation.

fixes https://github.com/docker/docker/issues/25483

**Note for reviewing**

I opened this PR for discussion, but as it is a change in both the CLI _and_ the API, I'm not sure what the best way forward is. I don't know if this can be seen as a regular bugfix, or if we need to start a deprecation traject.

ping @stevvooe @aluzzardi PTAL

thaJeztah committed:
Remove --name flag from service update

The --name flag was inadvertently added to
docker service update, but is not supported,
as it has various side-effects (e.g., existing
tasks are not renamed).

This removes the flag from the service update
command.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

aaronlehmann commented on vendor/src/github.com/docker/swarmkit/manager/controlapi/service.go:
```diff
@@ -18,6 +18,7 @@ import (
 
 var (
 	errNetworkUpdateNotSupported = errors.New("changing network in service is not supported")
+	errRenameNotSupported        = errors.New("service renaming is not supported")
```
nit: I prefer "renaming services is not supported"

thaJeztah replied:
yes, that may be better (I based this one on "service mode change is not allowed" below)

thaJeztah replied:
Opened https://github.com/docker/swarmkit/pull/1585

aaronlehmann commented on vendor/src/github.com/docker/swarmkit/manager/controlapi/service.go:
```diff
@@ -349,6 +350,11 @@ func (s *Server) UpdateService(ctx context.Context, request *api.UpdateServiceRe
 			return errModeChangeNotAllowed
 		}
+
+		if request.Spec != nil && service.Spec.Annotations.Name != request.Spec.Annotations.Name {
```
The `request.Spec != nil` check is not necessary, and we can also remove the others above. `validateServiceSpec` ensures that it is not `nil`.

thaJeztah replied:
Nice! Let me open a separate PR for the others in the SwarmKit repo

thaJeztah replied:
Opened https://github.com/docker/swarmkit/pull/1585

aaronlehmann commented on docs/reference/api/docker_remote_api_v1.24.md:
```diff
@@ -4785,7 +4785,6 @@ image](#create-an-image) section for more details.
     Content-Type: application/json
 
     {
-      "Name": "top",
```
`Name` is still part of the service update API - in fact, I believe that omitting it will trigger the service renaming check that you are added here. It is just imperative that the client not change the name from its previous value.

thaJeztah replied:
I tried doing an API request that omitted the "Name", and I think it worked, but I'll double check with the current code (as I made some modifications since, and possibly I had `request.Spec.Annotations.Name != nil && ...` in the previous version).

thaJeztah replied:
And, yes, I was wondering if there should be a separate `ServiceUpdate` API type (without the "Name"). IIUC, the overall design is that `Update` request should only have to send those parts of the Service / Spec that need to be _modified_ (only a "diff"), correct?

aaronlehmann replied:
No, the model is that the client should send the entire spec on a update.

See https://github.com/docker/swarmkit/issues/1394#issuecomment-240850719

thaJeztah commented:
We discussed this in the maintainers meeting, and this is a bugfix, so we can go ahead with this change.

@tonistiigi can you add the missing info on the API change? :heart:

stevvooe commented:
Design LGTM

The messaging is going to be complex here.

tonistiigi commented:
@thaJeztah As @aaronlehmann already commented, Docker side API should not change with this. We already send full object, even when no cli changes and we should continue to do so. We can add documentation that even though this field is part of the object you may not change that in a later update. We already have couple of restrictions like this.

thaJeztah commented:
@aaronlehmann @tonistiigi updated; added the "name" field back to the API, and updated the description to explain that updating the value is not allowed

aaronlehmann commented:
LGTM

tonistiigi commented:
LGTM

vieux commented:
LGTM

vieux commented:
@thaJeztah could you open the `vendor/src/github.com/docker/swarmkit/manager/controlapi/service.go` changes in swarmkit ?

thaJeztah commented:
opened https://github.com/docker/swarmkit/pull/1606

xiaods commented:
any udpate?

thaJeztah commented:
Rebased; SwarmKit changes are part of https://github.com/docker/docker/pull/27784/files#diff-bcfd894ce32b9a686a1ba4389bebad24R446

aluzzardi commented:
LGTM

vdemeester commented:
docs LGTM 🐮

thaJeztah commented:
author LGTM
